### PR TITLE
Adding in server, fault codec, tests, and examples

### PR DESCRIPTION
**Pull request by TimWhiting**
Here is the initial implementation. I welcome feedback, and I know there probably still needs to be some more testing and examples (especially of the server specific things).

### CODE REVIEW

**Review by a14n — COMMENTED** [review]
Thanks for your PR!

Regarding the server part I'd like to have a stronger separation between request handling and XML-RPC handling.

```dart
// in server.dart
class XmlRpcHandler {
  XmlRpcHandler({@required this.methods, this.codecs = standardCodecs});
  final Map<String, Function> methods;
  final List<Codec> codecs;
  // TODO in another PR to make review easier 
  bool acceptIntrospection;
  Future<XmlDocument> handle(XmlDocument xmlrpcCall) { ... }
}

// in simple_server.dart
import 'dart:io';
class XmlRpcServer {
  int port;
  int host:
...
}
```

Thus one can use shelf or something else to handle http calls.

**Comment by TimWhiting** [comment]
@a14n 
I believe I addressed all of your requests, reorganizing the classes and leaving out the introspection to make this PR smaller. I also fixed a few issues that I ran into when testing faults in the new example I added, and updated documentation to match.

File: lib/src/server.dart [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  /// The function registry
+  final Map<String, Function> methods;
+
+  /// Creates a [XmlRpcHandler] with the set of [codecs] for encoding and decoding
+  XmlRpcHandler({@required this.methods, this.codecs}) {
+    codecs ??= standardCodecs;
+  }
+
+  /// Marshalls the [data] from XML to Dart types, and then dispatches the function, and marshals the return value back into the XMLRPC format
+  XmlDocument handle(XmlDocument document) {
```
Comment (a14n) [inline lib/src/server.dart]: The return type should be `Future<XmlDocument>` to allow asynchronous operations.

File: lib/src/server.dart [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  /// The [codecs] used for encoding and decoding
+  List<Codec> codecs;
+
+  /// The function registry
+  final Map<String, Function> methods;
+
+  /// Creates a [XmlRpcHandler] with the set of [codecs] for encoding and decoding
+  XmlRpcHandler({@required this.methods, this.codecs}) {
+    codecs ??= standardCodecs;
+  }
```
Comment (a14n) [inline lib/src/server.dart]: We can use initializer list to have final on `codecs` field.

File: lib/src/server.dart [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    return XmlDocument([
+      XmlProcessing('xml', 'version="1.0"'),
+      XmlElement(XmlName('methodResponse'), [], methodResponseChildren)
+    ]);
+  }
+
+  /// Registers a [function] to be dispatched when recieving an XMLRPC call on a function with the name [name]
+  void registerFunction(Function function, String name) {
+    methods[name] = function;
+  }
```
Comment (a14n) [inline lib/src/server.dart]: Not sure it's worth keeping this method.

File: lib/src/server.dart [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      if (methods[method] != null) {
+        return Function.apply(methods[method], params);
+      } else {
+        return Fault(
+            2, 'No method by the name $method, registered with this server');
+      }
+    } on Exception {
+      return Fault(1, 'Dispatching $method, with params $params failed');
+    }
+  }
```
Comment (a14n) [inline lib/src/server.dart]: I'm not sure it's worth having those error codes hard-coded here.

I'd prefer to `assert(methods.contains(method))` at the begining of `handle` and let the caller deals with unexisting methods. This behaviour doesn't look specified and I also see some other codes (see https://stackoverflow.com/a/2045130/634863)

File: lib/src/simple_server.dart [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    } else {
+      _sendResponse(request, 501, 'Encoding $headerEncoding not supported');
+      return null;
+    }
+  }
+}
+
+/// Sends a simple Http response to the [request] with the specified [code] and [messsage]
+void _sendResponse(HttpRequest request, int code, String message) {
+  request.response.statusCode = 404;
```
Comment (a14n) [inline lib/src/simple_server.dart]:
```suggestion
  request.response.statusCode = code;
```

File: lib/server.dart [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+///     main() {
+///       final url = '...';
+///       xml_rpc
+///           .call(url, 'examples.getStateName', [41])
+///           .then((result) => print(result))
+///           .catchError((error) => print(error));
+///     }
+
+export 'src/server.dart';
+export 'src/simple_server.dart';
```
Comment (a14n) [inline lib/server.dart]: Could you move `src/server.dart` and `src/simple_server.dart` into `lib/`? The goal is to be able to use XmlRpcHandler without having `dart:io` imported.

In `lib/simple_server.dart` no problem to have `export 'server.dart';`.

File: lib/src/server.dart [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+          methodCall.findElements('params').first.findElements('param');
+      for (final arg in args) {
+        parsedArgs.add(
+          decode(getValueContent(arg.findElements('value').first), codecs),
+        );
+      }
+    } on StateError {
+      // No arguments given
+    }
+    final returnValue = _dispatch(methodName, parsedArgs);
```
Comment (a14n) [inline lib/src/server.dart]: We can avoid try-catch with null-safe operators.

```suggestion
    final args = methodCall
        .findElements('params')
        ?.first
        ?.findElements('param')
        ?.map((param) => param.findElements('value').first)
        ?.map(getValueContent)
        ?.map((e) => decode(e, codecs))
        ?.toList();
    final returnValue = await Function.apply(methods[methodName], args ?? []);
```

File: lib/src/simple_server.dart [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  /// before making any client requests
+  XmlRpcServer({
+    @required this.host,
+    @required this.port,
+    @required this.requestHandler,
+    this.encoding = utf8,
+  });
+
+  /// Starts up the [XmlRpcServer] and starts listening to requests
+  Future<void> serveForever();
```
Comment (a14n) [inline lib/src/simple_server.dart]: `start` instead of `serveForever` ?

File: lib/src/simple_server.dart [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  /// Creates a [XmlRpcServer] that will bind to the specified [host] and [port]
+  ///
+  /// as well as the String [encoding] for http requests and responses
+  ///
+  /// You must await the call to [serverForever] to start up the server
+  /// before making any client requests
+  XmlRpcServer({
+    @required this.host,
+    @required this.port,
+    @required this.requestHandler,
```
Comment (a14n) [inline lib/src/simple_server.dart]: rename to `handler`

**Review by a14n — COMMENTED** [review]
Thanks for the update and for your time. Here's another bunch of comments.

File: lib/src/server.dart (hunk 4 above)
Comment (TimWhiting) [reply lib/src/server.dart]: Yeah, according to the xmlrpc spec the error codes are supposed to be application specific. If the caller is in charge of dealing with non-existing methods, we still need to send back a fault, and protect the server from crashing because of a malicious client. Won't the assert get taken out when not in debug, crashing the server? Maybe we can have fields in the XmlRpcHandler class that the user can set these codes.

File: lib/server.dart (hunk 6 above)
Comment (TimWhiting) [reply lib/server.dart]: Okay, I did that, though I don't think you can spin up a webserver in non dart:io contexts anyways.

File: lib/src/server.dart (hunk 7 above)
Comment (TimWhiting) [reply lib/src/server.dart]: If there are no params, `methodCall.findElements('params')` does not return `null`, but the getter `first` throws a `StateError`. I would need to check if `methodCall.findElements('params').isEmpty()` which doesn't fit into the null cascade as easily. But I think I can do something similar, and avoid the try / catch.

File: lib/src/server.dart (hunk 4 above)
Comment (TimWhiting) [reply lib/src/server.dart]: I think I have a good solution for this that allows the user to specify the fault code for the missing method, and a custom fault for any fault from a method failure.

**Comment by TimWhiting** [comment]
@a14n 
Ready for feedback, whenever you have time to review.

File: lib/server.dart [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    List<Codec> codecs,
+    MethodFailureSignature methodFailureHandler,
+    this.methodFailureCode,
+    this.noExistingMethodCode,
+  }) : codecs = codecs ?? standardCodecs {
+    this.methodFailureHandler =
+        methodFailureHandler ?? defaultMethodFailureHandler;
+  }
+
+  Fault defaultMethodFailureHandler(
```
Comment (a14n) [inline lib/server.dart]: we should make it static private by using a hard-coded code (like `-32500` - from http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php linked in https://stackoverflow.com/a/2045130/634863)

File: lib/server.dart [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+/// Exception or error that occurred
+typedef MethodFailureSignature = Fault Function(String, List<dynamic>, dynamic);
+
+/// A [XmlRpcHandler] handles the handling RPC functions along with marshalling the arguments and results to / from XMLRPC spec
+///
+/// Has a set of [codecs] for encoding and decoding the datatypes
+/// To register a method or function with the dispatcher call [registerFunction] along with the function and name
+class XmlRpcHandler {
+  /// The [codecs] used for encoding and decoding
+  List<Codec> codecs;
```
Comment (a14n) [inline lib/server.dart]: final

File: lib/server.dart [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+  /// The function registry
+  final Map<String, Function> methods;
+
+  /// A function that gets called on a method failure with the following data
+  ///
+  /// Name of the method being called
+  /// Parameters that were passed in
+  /// Exception or error that occurred
+  MethodFailureSignature methodFailureHandler;
```
Comment (a14n) [inline lib/server.dart]: could be made final like codecs

File: lib/server.dart [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  /// The error code for a missing method
+  final int noExistingMethodCode;
+
+  /// Creates a [XmlRpcHandler] with the set of [codecs] for encoding and decoding
+  XmlRpcHandler({
+    @required this.methods,
+    List<Codec> codecs,
+    MethodFailureSignature methodFailureHandler,
+    this.methodFailureCode,
+    this.noExistingMethodCode,
```
Comment (a14n) [inline lib/server.dart]: we should make this param required and assert non-null (or put a default value like `-32601` - from http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php linked in https://stackoverflow.com/a/2045130/634863)

File: lib/server.dart [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+  /// The error code for a missing method
+  final int noExistingMethodCode;
+
+  /// Creates a [XmlRpcHandler] with the set of [codecs] for encoding and decoding
+  XmlRpcHandler({
+    @required this.methods,
+    List<Codec> codecs,
+    MethodFailureSignature methodFailureHandler,
+    this.methodFailureCode,
```
Comment (a14n) [inline lib/server.dart]: Having both `methodFailureHandler` and `methodFailureCode` looks redundant. I'd remove `methodFailureCode`.

File: lib/server.dart [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  Future<XmlDocument> handle(XmlDocument document) async {
+    final methodCall = document.findElements('methodCall').first;
+    final methodName = methodCall.findElements('methodName').first.text;
+    var returnValue;
+    if (methods.containsKey(methodName)) {
+      final params = methodCall.findElements('params');
+      final parsedArgs = params.isNotEmpty
+          ? params.first
+              .findElements('param')
+              .map((arg) => arg.findElements('value').first)
```
Comment (a14n) [inline lib/server.dart]: Thanks for the reminder that `.first` can throw if list is empty. I'm sorry for the unnecessary change I requested.

Here a StateError could be thrown if no `<value>`...

Moreover decoders can throw ArgumentError ; we should try-catch those decoding/argument issues.

**Review by a14n — COMMENTED** [review]
Some additional comments.

I like how error/fault are handled now. The code looks fine.

File: lib/server.dart (hunk 15 above)
Comment (TimWhiting) [reply lib/server.dart]: Also, I'm adding handling of encoding errors

**Comment by TimWhiting** [comment]
@a14n 
I believe I addressed all your comments and added in better exception handling for decoding / encoding, and malformed xml errors.

File: lib/server.dart [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+            : [];
+
+        returnValue = await _dispatch(methodName, parsedArgs);
+      } else {
+        returnValue = Fault(noExistingMethodCode,
+            'Server error. No method by the name $methodName, registered with this server');
+      }
+    } on ArgumentError catch (e) {
+      returnValue = Fault(unsupportedEncoding,
+          'Parse error. Decoding arguments failed, got: ${document.toXmlString()}, error was $e');
```
Comment (a14n) [inline lib/server.dart:95]: `unsupportedEncoding` is more related to char encoding (utf8...) imho.

The more I think about how to deal with those error cases the more I tend to have custom exceptions thrown and let the implementation handle them. Something like:

```dart
class XmlRpcException extends Exception {}
class XmlRpcRequestFormatException extends XmlRpcException {}
/// When an exception occurs in call
class XmlRpcCallException extends XmlRpcException {}
/// When an exception occurs in response encoding
class XmlRpcResponseEncodingException extends XmlRpcException {}
```

We could also add an helper method on `XmlRpcHandler` to encode `Fault`. This will allow to get the xml to be sent to the client.

Thus in `simple_server` we would have:

```dart
        try {
          response = (await handler.handle(parse(result))).toXmlString();
        } on XmlRpcRequestFormatException {
          response = handler.encodeFault(Fault(-32700, 'parse error. not well formed')).toXmlString();
        } on XmlRpcCallException {
          response = handler.encodeFault(Fault(-32500, 'application error')).toXmlString();
        }
```

WDYT?

**Review by a14n — COMMENTED** [review]
Looks good.

Let me know if you don't have more time to spend on this PR. I can merge it and make some changes afterwards.

File: lib/server.dart (hunk 16 above)
Comment (TimWhiting) [reply lib/server.dart:95]: Yeah, I think that is a better approach.

If you have the time, go ahead and merge it in and make the changes, since I think you have a better idea on how you would like to expose this to the end user.

If you are busy, I'd be willing to make the adjustments from your comments.

Thanks!

**Comment by a14n** [comment]
I made some refactoring on master. It would be great if you could test it on your side (if you already have a xmlrpc server). If it's ok for you I will push a new version to pub (after adding some tests).

**Comment by TimWhiting** [comment]
Sounds good, looks like you already published it. I don't currently have an xmlrpc server to test it on except the simple examples in the examples directory (which seem to be working fine). I needed this pull request in order to progress on the server I'm working on as part of a larger project. I'll let you know in a few weeks if I find any issues while working on that server.

Thanks,
It was a pleasure working with you
